redis/server.cc

```diff
@@ -0,0 +1,259 @@
+    return futurize_apply([this, request = std::move(request), &opts, permit] () mutable {
+        return _server._query_processor.local().process(std::move(request), seastar::ref(opts), permit).then([] (auto&& message) {
+            return make_ready_future<redis_server::result> (std::move(message));
+        });
+    });
+}
+
+redis_server::connection::connection(redis_server& server, socket_address server_addr, connected_socket&& fd, socket_address addr)
+    : _server(server)
+    , _server_addr(server_addr)
+    , _fd(std::move(fd))
+    , _read_buf(_fd.input())
+    , _write_buf(_fd.output())
+    , _options(server._config._read_consistency_level, server._config._write_consistency_level, server._config._timeout_config, server._auth_service, addr)
+{
+    ++_server._total_connections;
+    ++_server._current_connections;
+    _server._connections_list.push_back(*this);
+}
+
+redis_server::connection::~connection() {
+    --_server._current_connections;
+    _server._connections_list.erase(_server._connections_list.iterator_to(*this));
+    _server.maybe_idle();
+}
+
+future<> redis_server::connection::process()
+{
+    return do_until([this] {
+        return _read_buf.eof();
+    }, [this] {
+        return with_gate(_pending_requests_gate, [this] {
+            return process_request();
+        });
+    }).then_wrapped([this] (future<> f) {
+        try {
+            f.get();
+        }
+        catch (redis_exception& e) {
+        logging.info(" redis exception = {} ", e.what());
+            write_reply(e);
+        }
+        catch (std::exception& e) {
+        logging.info(" std::exception = {} ", e.what());
+            write_reply(redis_exception { e.what() });
+        }
+        catch (...) {
+        logging.info(" ... exception ");
+            write_reply(redis_exception { "Unknown exception" });
+        }
+    }).finally([this] {
+        return _pending_requests_gate.close().then([this] {
+            return _ready_to_respond.finally([this] {
+                return _write_buf.close();
+            });
+        });
+    });
+}
+
+future<> redis_server::connection::shutdown()
+{
+    try {
+        _fd.shutdown_input();
+        _fd.shutdown_output();
+    } catch (...) {
+    }
+    return make_ready_future<>();
+}
+
+thread_local redis_server::connection::execution_stage_type redis_server::connection::_process_request_stage {"redis_transport", &connection::process_request_one};
+
+future<redis_server::result> redis_server::connection::process_request_internal() {
+    return _process_request_stage(this, std::move(_parser.get_request()), seastar::ref(_options), empty_service_permit());
+}
+
+void redis_server::connection::write_reply(const redis_exception& e)
+{
+    _ready_to_respond = _ready_to_respond.then([this, exception_message = e.what_message()] () mutable {
+        return redis_message::make_exception(exception_message).then([this] (auto&& result) {
+            auto message = result.message();
```

Comment:
Why do you need to create a ready future only to extract the message from it? Or what are you trying to do here?